### ISSUE TITLE
Handle wrapping behavior

### ISSUE THREAD

**Issue by Schaeff** [powdr-org/powdr#300]
- #151 removes wrapping behavior from witness generation, because it is not a requirement in the halo2 backend due to extra rows needed for eg. blinding factors.
- On the flip side, it is a requirement in eSTARK, so even though witness generation is currently fine according to pilcom, it could fail if a witness does not satisfy the wrapping constraint.
- One idea would be to toggle wrapping in the executor depending on which backend is to be used, but this breaks the clear separation.
- This was moved to an issue as it was not clear at the time how to deal with it, and witness generation did not rely on the wrapping behavior in the existing state machines.

**Comment by chriseth** [comment]
Some more notes: We might already use wrapping to compute the values for the first row, for example if we have an identity of the form `x' = x + 1`, then it will actually be evaluated as `x = x[-1] + 1`.

We should analyze the PIL to check if it depends on wrapping or not (substitute fixed columns and see if next references are still there).

In my opinion, the cleanest solution would be to change the asm->pil generator to not rely on wrapping and report if there are "free" rows.

If pil is specified by the user, the system should warn that it is not compatible with halo2 and could even offer to introduce additional columns and factors to fix it (as we currently do).

**Comment by leonardoalt** [comment]
Is this fine now? @Schaeff @chriseth

**Comment by chriseth** [comment]
I think we are still just ignoring the problem and thus introducing nondeterminism with halo2.

**Comment by leonardoalt** [comment]
I don't think so. In halo2 we explicitly add 2 selector columns, `enable_cur` and `enable_next`, that guard the used rows and the used rows - 1, respectively. I wouldn't say this introduces nondeterminism. This would be true only for machines that rely on the cyclic behavior, which I don't think we have.

**Comment by leonardoalt** [comment]
This is sound now in Halo2.
